pennylane/ops/qubit.py

```diff
@@ -2146,17 +2146,29 @@ class QubitUnitary(Operation):
     par_domain = "A"
     grad_method = None
 
-    @classmethod
-    def _matrix(cls, *params):
-        U = np.asarray(params[0])
+    def __init__(self, *params, wires, do_queue=True):
+        wires = Wires(wires)
 
-        if U.ndim != 2 or U.shape[0] != U.shape[1]:
-            raise ValueError("Operator must be a square matrix.")
+        # For pure QubitUnitary operations (not controlled), check that the number
+        # of wires fits the dimensions of the matrix
+        if not isinstance(self, ControlledQubitUnitary):
+            U = np.asarray(params[0])
```

Comment:
I realize that the previous version of the code does this, _but_, I only just noticed it 😆 

We can't assume that `params[0]` will be a NumPy array - it could very easily be a TensorFlow or a PyTorch tensor. As a result, best to use `qml.math` here. E.g.,

```python
U = params[0]
shape = qml.math.shape(U)
```

(and then down below, `qml.math.dot` and `qml.math.conj`)